libp2p/protocols/secure/noise.nim

```diff
@@ -0,0 +1,557 @@
+##  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+## at your option.
+## This file may not be copied, modified, or distributed except according to
+## those terms.
+
+import chronos
+import chronicles
+import random
+import stew/[endians2, byteutils]
+import nimcrypto/[utils, sysrand, sha2, hmac]
+import ../../connection
+import ../../peer
+import ../../peerinfo
+import ../../protobuf/minprotobuf
+import secure,
+       ../../crypto/[crypto, chacha20poly1305, curve25519, hkdf],
+       ../../stream/bufferstream
+
+logScope:
+  topic = "Noise"
+
+const
+  # https://godoc.org/github.com/libp2p/go-libp2p-noise#pkg-constants
+  NoiseCodec* = "/noise"
+  
+  PayloadString = "noise-libp2p-static-key:"
+
+  ProtocolXXName = "Noise_XX_25519_ChaChaPoly_SHA256"
+
+  # Empty is a special value which indicates k has not yet been initialized.
+  EmptyKey: ChaChaPolyKey = [0.byte, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
+  NonceMax = uint64.high - 1 # max is reserved
+
+type
+  KeyPair = object
+    privateKey: Curve25519Key
+    publicKey: Curve25519Key
+  
+  # https://noiseprotocol.org/noise.html#the-cipherstate-object
+  CipherState = object
+    k: ChaChaPolyKey
+    n: uint64
+
+  # https://noiseprotocol.org/noise.html#the-symmetricstate-object
+  SymmetricState = object
+    cs: CipherState
+    ck: ChaChaPolyKey
+    h: MDigest[256]
+
+  # https://noiseprotocol.org/noise.html#the-handshakestate-object
+  HandshakeState = object
+    ss: SymmetricState
+    s: KeyPair
+    e: KeyPair
+    rs: Curve25519Key
+    re: Curve25519Key
+
+  HandshakeResult = object
+    cs1: CipherState
+    cs2: CipherState
+    remoteP2psecret: seq[byte]
+    rs: Curve25519Key
+  
+  Noise* = ref object of Secure
+    localPrivateKey: PrivateKey
+    localPublicKey: PublicKey
+    noisePrivateKey: Curve25519Key
+    noisePublicKey: Curve25519Key
+    commonPrologue: seq[byte]
+
+  NoiseConnection* = ref object of SecureConnection
+    readCs: CipherState
+    writeCs: CipherState
+
+  NoiseHandshakeError* = object of CatchableError
+  NoiseDecryptTagError* = object of CatchableError
+  NoiseOversizedPayloadError* = object of CatchableError
+  NoiseNonceMaxError* = object of CatchableError # drop connection on purpose
+
+# Utility
+
+proc genKeyPair(): KeyPair =
+  result.privateKey = Curve25519Key.random()
+  result.publicKey = result.privateKey.public()
+    
+proc hashProtocol(name: string): MDigest[256] =
+  # If protocol_name is less than or equal to HASHLEN bytes in length,
+  # sets h equal to protocol_name with zero bytes appended to make HASHLEN bytes.
+  # Otherwise sets h = HASH(protocol_name).
+
+  if name.len <= 32:
+    copyMem(addr result.data[0], unsafeAddr name[0], name.len)
+  else:
+    result = sha256.digest(name)
+
+proc dh(priv: Curve25519Key, pub: Curve25519Key): Curve25519Key =
+  Curve25519.mul(result, pub, priv)
+
+# Cipherstate
+
+proc init(_: type[CipherState]; key: ChaChaPolyKey): CipherState =
```

Comment:
So here, `CipherState(k: key, n: 0)` is also nicer - no `result` repetition, no partially initialized `CipherState` ever exists (abstractly) so no risk of copy-paste problems or partial init